pkg/xds/bootstrap/tracing.go

```diff
@@ -0,0 +1,127 @@
+	envoy_api "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	envoy_api_v2_core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	envoy_api_v2_endpoint "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
+	envoy_bootstrap "github.com/envoyproxy/go-control-plane/envoy/config/bootstrap/v2"
+	envoy_config_trace_v2 "github.com/envoyproxy/go-control-plane/envoy/config/trace/v2"
+)
+
+func AddTracingConfig(bootstrap *envoy_bootstrap.Bootstrap, backend *mesh_proto.TracingBackend) error {
+	if zipkin, ok := backend.GetType().(*mesh_proto.TracingBackend_Zipkin_); ok {
+		cluster, tracingCfg, err := zipkinConfig(bootstrap, zipkin.Zipkin, backend.Name)
+		if err != nil {
+			return err
+		}
+		if bootstrap.StaticResources == nil {
+			bootstrap.StaticResources = &envoy_bootstrap.Bootstrap_StaticResources{}
+		}
+		bootstrap.StaticResources.Clusters = append(bootstrap.StaticResources.Clusters, cluster)
+		bootstrap.Tracing = tracingCfg
+	}
+	return nil
+}
+
+func zipkinConfig(bootstrap *envoy_bootstrap.Bootstrap, zipkin *mesh_proto.TracingBackend_Zipkin, backendName string) (*envoy_api.Cluster, *envoy_config_trace_v2.Tracing, error) {
+	url, err := net_url.ParseRequestURI(zipkin.Url)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "invalid URL of Zipkin")
+	}
+
+	cluster, err := zipkinCluster(backendName, url)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	zipkinConfig := envoy_config_trace_v2.ZipkinConfig{
+		CollectorCluster:         cluster.Name,
+		CollectorEndpoint:        url.Path,
+		TraceId_128Bit:           zipkin.TraceId128Bit,
+		CollectorEndpointVersion: apiVersion(zipkin, url),
+	}
+	zipkinConfigAny, err := ptypes.MarshalAny(&zipkinConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+	tracingConfig := &envoy_config_trace_v2.Tracing{
+		Http: &envoy_config_trace_v2.Tracing_Http{
+			Name: "envoy.zipkin",
+			ConfigType: &envoy_config_trace_v2.Tracing_Http_TypedConfig{
+				TypedConfig: zipkinConfigAny,
+			},
+		},
+	}
+	return cluster, tracingConfig, nil
+}
+
+func apiVersion(zipkin *mesh_proto.TracingBackend_Zipkin, url *net_url.URL) envoy_config_trace_v2.ZipkinConfig_CollectorEndpointVersion {
+	if zipkin.ApiVersion == "" { // try to infer it from the URL
+		if url.Path == "/api/v1/spans" {
+			return envoy_config_trace_v2.ZipkinConfig_HTTP_JSON_V1
+		} else if url.Path == "/api/v2/spans" {
+			return envoy_config_trace_v2.ZipkinConfig_HTTP_JSON
+		}
+	} else {
+		switch zipkin.ApiVersion {
+		case "httpJsonV1":
+			return envoy_config_trace_v2.ZipkinConfig_HTTP_JSON_V1
+		case "httpJson":
+			return envoy_config_trace_v2.ZipkinConfig_HTTP_JSON
+		case "httpProto":
+			return envoy_config_trace_v2.ZipkinConfig_HTTP_PROTO
+		}
+	}
+	return envoy_config_trace_v2.ZipkinConfig_HTTP_JSON
+}
+
+const zipkinClusterTimeout = 10 * time.Second
+
+func zipkinCluster(backendName string, url *net_url.URL) (*envoy_api.Cluster, error) {
+	port, err := strconv.Atoi(url.Port())
```

Comment:
This will result in the error if URL has no port.

If the port is absolutely mandatory, should we update `Mesh` validator then ?